src/python/grpcio_status/grpc_status/rpc_status.py

```diff
@@ -0,0 +1,60 @@
+# Copyright 2018 The gRPC Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Reference implementation for status mapping in gRPC Python."""
+
+import grpc
+
+# Have to import google.protobuf before google.rpc, otherwise Python will fail
+#   to locate module rpc under google.
+import google.protobuf  # pylint: disable=unused-import
+from google.rpc import status_pb2
+
+_CODE_TO_GRPC_CODE_MAPPING = dict([(x.value[0], x) for x in grpc.StatusCode])
+_GRPC_CODE_TO_CODE_MAPPING = dict([(x, x.value[0]) for x in grpc.StatusCode])
+
+_GRPC_DETAILS_METADATA_KEY = 'grpc-status-details-bin'
+
+
+def from_call(call):
+    """Returns a google.rpc.status.Status message corresponding to a given call.
+
+    Args:
+      call: A grpc.Call instance representing an RPC.
+
+    Returns:
+      A google.rpc.status.Status message representing the status of the RPC.
+    """
+    for metadatum in call.trailing_metadata():
+        if metadatum.key == _GRPC_DETAILS_METADATA_KEY:
+            return status_pb2.Status.FromString(metadatum.value)
+    return status_pb2.Status(
+        code=_GRPC_CODE_TO_CODE_MAPPING[call.code()], message=call.details())
+
+
+def abort_with_status(servicer_context, status):
+    """Aborts an RPC on the service side with the specified status message.
+
+    Args:
+      servicer_context: A grpc.ServicerContext representing the RPC to abort.
+      status: a google.rpc.status.Status message representing the non-OK status
+      to terminate the RPC with and communicate it to the client.
+
+    Raises:
+      Exception: An exception is always raised to signal the abortion the
+        RPC to the gRPC runtime.
+    """
+    servicer_context.set_trailing_metadata(((_GRPC_DETAILS_METADATA_KEY,
+                                             status.SerializeToString()),))
+    servicer_context.abort(_CODE_TO_GRPC_CODE_MAPPING[status.code],
```

Comment:
The `status_pb2.Status` instance's code may not be a valid gRPC code. As is, we would need to handle this and probably still abort the call...this is another reason it would be nice to decouple (conversion from `status_pb2.Status` to gRPC status) and the action of actually aborting the RPC.

Reply:
I added a unit test for invalid status code. The translation will raise `ValueError` if the code failed to map to gRPC's `StatusCode`.